R/Spectra-functions.R

```diff
@@ -437,3 +433,55 @@ combineSpectra <- function(x, f = x$dataStorage, p = x$dataStorage,
                     MoreArgs = list(FUN = FUN, ...), BPPARAM = BPPARAM)
     .concatenate_spectra(res)
 }
+
+#' @description
+#'
+#' Internal function to check if any (or all) of the provided `mz` values are
+#' in the spectras' m/z.
+#'
+#' @param x `Spectra` object
+#'
+#' @param mz `numeric` of m/z value(s) to check in each spectrum of `x`.
+#'
+#' @param tolarance `numeric(1)` with the tolerance.
+#'
+#' @param ppm `numeric(1)` with the ppm.
+#'
+#' @param condFun `function` such as `any` or `all`.
+#'
+#' @param parallel `BiocParallel` parameter object.
+#'
+#' @return `logical` same length than `x`.
+#'
+#' @author Johannes Rainer
+#'
+#' @importFrom MsCoreUtils common
+#'
+#' @noRd
+.has_mz <- function(x, mz = numeric(), tolerance = 0, ppm = 20, condFun = any,
+                    parallel = SerialParam()) {
+    mzs <- mz(x, BPPARAM = parallel)
+    vapply(mzs, FUN = function(z) {
+        condFun(common(mz, z, tolerance = tolerance, ppm = ppm))
+    }, logical(1))
+}
+
+#' @description
+#'
+#' Same as `.has_mz` only that a different `mz` is used for each spectrum in
+#' `x`. Length of `mz` is thus expected to be equal to length of `x`.
+#'
+#' @param mz `numeric` **same length than `x`**.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.has_mz_each <- function(x, mz = numeric(), tolerance = 0, ppm = 20,
+                         parallel = SerialParam()) {
+    mzs <- mz(x, BPPARAM = parallel)
+    mapply(mzs, mz, FUN = function(z, y) {
+        if (is.na(y))
+            NA # if m/z is NA is better to return NA instead of FALSE
+        else common(y, z, tolerance = tolerance, ppm = ppm)
```

Comment:
I'm also not sure. `NA == 4` returns `NA` while `NA %in% c(3, 4)` returns `FALSE`. So, there seems also to be no real consistency in R. My reasoning was more that `precursorMz` will be `NA` for MS1 spectra and for these it *might* be better if we returned `NA`. But I would also be OK with `FALSE`. What would you prefer?

Reply:
Before merging I would like to get your thoughts on that @sgibb

Reply:
I would prefer `NA` for the MS1 as well. As `closest` is intended to be as similar as possible to `match` it would be useful to keep `common` similar to `%in%` but the inconsistency is annoying.